// es-abstract/operations/2017.js
'use strict';
module.exports = {
	IsPropertyDescriptor: 'https:
	IsAccessorDescriptor: 'https:
	IsDataDescriptor: 'https:
	IsGenericDescriptor: 'https:
	FromPropertyDescriptor: 'https:
	ToPropertyDescriptor: 'https:
	CompletePropertyDescriptor: 'https:
	ToPrimitive: 'https:
	ToBoolean: 'https:
	ToNumber: 'https:
	ToInteger: 'https:
	ToInt32: 'https:
	ToUint32: 'https:
	ToInt16: 'https:
	ToUint16: 'https:
	ToInt8: 'https:
	ToUint8: 'https:
	ToUint8Clamp: 'https:
	ToString: 'https:
	ToObject: 'https:
	ToPropertyKey: 'https:
	ToLength: 'https:
	CanonicalNumericIndexString: 'https:
	ToIndex: 'https:
	RequireObjectCoercible: 'https:
	IsArray: 'https:
	IsCallable: 'https:
	IsConstructor: 'https:
	IsExtensible: 'https:
	IsInteger: 'https:
	IsPropertyKey: 'https:
	IsRegExp: 'https:
	SameValue: 'https:
	SameValueZero: 'https:
	SameValueNonNumber: 'https:
	Get: 'https:
	GetV: 'https:
	Set: 'https:
	CreateDataProperty: 'https:
	CreateMethodProperty: 'https:
	CreateDataPropertyOrThrow: 'https:
	DefinePropertyOrThrow: 'https:
	DeletePropertyOrThrow: 'https:
	GetMethod: 'https:
	HasProperty: 'https:
	HasOwnProperty: 'https:
	Call: 'https:
	Construct: 'https:
	SetIntegrityLevel: 'https:
	TestIntegrityLevel: 'https:
	CreateArrayFromList: 'https:
	CreateListFromArrayLike: 'https:
	Invoke: 'https:
	OrdinaryHasInstance: 'https:
	SpeciesConstructor: 'https:
	EnumerableOwnProperties: 'https:
	GetIterator: 'https:
	IteratorNext: 'https:
	IteratorComplete: 'https:
	IteratorValue: 'https:
	IteratorStep: 'https:
	IteratorClose: 'https:
	CreateIterResultObject: 'https:
	CreateListIterator: 'https:
	Type: 'https:
	thisNumberValue: 'https:
	thisTimeValue: 'https:
	thisStringValue: 'https:
	RegExpExec: 'https:
	RegExpBuiltinExec: 'https:
	IsConcatSpreadable: 'https:
	IsPromise: 'https:
	ArraySpeciesCreate: 'https:
	ObjectCreate: 'https:
	AdvanceStringIndex: 'https:
	OrdinarySet: 'https:
	NormalCompletion: 'https:
	IsSharedArrayBuffer: 'https:
};
